firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/UpdateApkClient.java

```diff
@@ -47,57 +45,52 @@
     this.downloadExecutor = Executors.newSingleThreadExecutor();
-    this.downloadHandler = HandlerCompat.createAsync(Looper.getMainLooper());
     this.firebaseApp = firebaseApp;
   }
 
   public void updateApk(
-      @NonNull String downloadUrl,
-      @NonNull Activity currentActivity,
       @NonNull UpdateTaskImpl updateTask,
-      @NonNull TaskCompletionSource updateAppTaskCompletionSource) {
-    this.updateTask = updateTask;
-    this.updateAppTaskCompletionSource = updateAppTaskCompletionSource;
-    downloadApk(downloadUrl)
+      @NonNull String downloadUrl,
+      @NonNull Activity currentActivity) {
+
+    downloadApk(downloadUrl, updateTask)
         .addOnSuccessListener(
+            downloadExecutor,
             file ->
                 install(file.getPath(), currentActivity)
                     .addOnSuccessListener(
-                        downloadExecutor,
-                        Void -> {
-                          updateTask.updateProgress(
-                              UpdateProgress.builder()
-                                  .setApkFileTotalBytes(file.length())
-                                  .setApkBytesDownloaded(file.length())
-                                  .setUpdateStatus(UpdateStatus.DOWNLOADED)
-                                  .build());
-                          updateAppTaskCompletionSource.setResult(null);
+                        unused -> {
+                          postUpdateProgress(
+                              updateTask, file.length(), file.length(), UpdateStatus.DOWNLOADED);
+                          updateTask.setResult();
                         })
                     .addOnFailureListener(
                         e ->
-                            setUpdateAppErrorWithDefault(
+                            setTaskCompletionErrorWithDefault(
+                                updateTask,
                                 e,
                                 new FirebaseAppDistributionException(
                                     Constants.ErrorMessages.NETWORK_ERROR,
                                     FirebaseAppDistributionException.Status.INSTALLATION_FAILURE))))
         .addOnFailureListener(
+            downloadExecutor,
             e ->
-                setUpdateAppErrorWithDefault(
+                setTaskCompletionErrorWithDefault(
+                    updateTask,
                     e,
                     new FirebaseAppDistributionException(
                         Constants.ErrorMessages.NETWORK_ERROR,
                         FirebaseAppDistributionException.Status.DOWNLOAD_FAILURE)));
   }
 
-  private @NonNull Task<File> downloadApk(@NonNull String downloadUrl) {
+  @VisibleForTesting
+  @NonNull
+  Task<File> downloadApk(@NonNull String downloadUrl, UpdateTaskImpl updateTask) {
     if (downloadTaskCompletionSource != null
         && !downloadTaskCompletionSource.getTask().isComplete()) {
       downloadCancellationTokenSource.cancel();
```

Comment:
Why do we cancel this task instead of returning the same one?

Reply:
I think it should be the same one to be consistent with the changes we made to the other completion source parts, I'll update the logic